sklearn/pipeline.py

```diff
@@ -177,20 +177,31 @@ def _validate_steps(self):
         for t in transformers:
             if t is None or t == 'passthrough':
                 continue
-            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
-                    hasattr(t, "transform")):
+            if (not hasattr(t, "fit_transform") or
+                    (not (hasattr(t, "fit") and hasattr(t, "transform")))):
                 raise TypeError("All intermediate steps should be "
-                                "transformers and implement fit and transform "
-                                "or be the string 'passthrough' "
-                                "'%s' (type %s) doesn't" % (t, type(t)))
+                                "transformers and implement fit and "
+                                "transform, fit_transform or be the string "
+                                "'passthrough'. '%s' (type %s) doesn't"
+                                % (t, type(t)))
+            elif not hasattr(t, "transform"):
+                warnings.warn("Intermediate step '%s' (type %s) does not have "
+                              "transform, pipeline is not reusable."
+                              % (t, type(t)))
 
         # We allow last estimator to be None as an identity transformation
         if (estimator is not None and estimator != 'passthrough'
-                and not hasattr(estimator, "fit")):
+                and not (hasattr(estimator, "fit") or
+                         hasattr(estimator, "fit_transform"))):
             raise TypeError(
-                "Last step of Pipeline should implement fit "
-                "or be the string 'passthrough'. "
+                "Last step of Pipeline should implement fit, "
+                "fit_transform or be the string 'passthrough'. "
                 "'%s' (type %s) doesn't" % (estimator, type(estimator)))
+        elif (not hasattr(estimator, "transform") and
+                  not hasattr(estimator, "predict")):
```

Comment:
You can split this into 4 lines (because of PEP8):
```python
elif (
    not hasattr(estimator, "transform") and
    not hasattr(estimator, "predict")
):
```
or rephrase it (putting `not` before the parenthesis and replacing `and` with `or`):
```python
elif not (hasattr(estimator, "transform") or
          hasattr(estimator, "predict")):

```

Reply:
The latter one looks nicer to me, will change it to that